src/nvme/linux.c

```diff
@@ -543,6 +559,187 @@ int nvme_gen_dhchap_key(char *hostnqn, enum nvme_hmac_alg hmac,
+			       unsigned char *retained_key, int key_len)
+{
+	key_serial_t key = -ENOKEY;
+	const char *key_type = "psk";
+	EVP_PKEY_CTX *ctx;
+	char *psk_identity;
+	unsigned char *psk;
+	size_t psk_len;
+	long int err;
+
+	psk_identity = malloc(strlen(hostnqn) + strlen(subsysnqn) + 12);
+	if (!psk_identity)
+		return ENOMEM;
+
+	sprintf(psk_identity, "NVMeR%02d %s %s", md == EVP_sha256() ? 1 : 2,
+		hostnqn, subsysnqn);
+
+	psk = malloc(key_len);
+	if (!psk) {
+		err = ENOMEM;
+		goto out_free_identity;
+	}
+	psk_len = key_len;
+
+	ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_HKDF, NULL);
+	if (!ctx) {
+		err = ENOMEM;
+		goto out_free_psk;
+	}
+
+	err = -ENOKEY;
+	if (EVP_PKEY_derive_init(ctx) <= 0)
+		goto out_free_ctx;
+	if (EVP_PKEY_CTX_set_hkdf_md(ctx, md) <= 0)
+		goto out_free_ctx;
+	if (EVP_PKEY_CTX_set1_hkdf_key(ctx, retained_key, key_len) <= 0)
+		goto out_free_ctx;
+	if (EVP_PKEY_CTX_add1_hkdf_info(ctx, "tls13 ", 6) <= 0)
+		goto out_free_ctx;
+	if (EVP_PKEY_CTX_add1_hkdf_info(ctx, "nvme-tls-psk", 12) <= 0)
+		goto out_free_ctx;
+	if (EVP_PKEY_CTX_add1_hkdf_info(ctx, psk_identity,
+					strlen(psk_identity)) <= 0)
+		goto out_free_ctx;
+
+	if (EVP_PKEY_derive(ctx, psk, &psk_len) <= 0) {
+		fprintf(stderr, "EVP_KDF_derive failed\n");
+	}
+
+	key = keyctl_search(keyring, key_type, psk_identity, 0);
+	if (key >= 0) {
+		printf("updating %s key '%s'\n",
+		       key_type, psk_identity);
+		err = keyctl_update(key, psk, psk_len);
+		if (err) {
+			fprintf(stderr, "updating %s key '%s' failed\n",
+				key_type, psk_identity);
+			key = err;
+		}
+	} else {
+		printf("adding %s key '%s'\n", key_type, psk_identity);
+		key = add_key(key_type, psk_identity,
+			      psk, psk_len, keyring);
+		if (key < 0)
+			fprintf(stderr, "adding %s key '%s' failed, error %d\n",
+				key_type, psk_identity, errno);
+	}
+
+out_free_ctx:
+	EVP_PKEY_CTX_free(ctx);
+out_free_psk:
+	free(psk);
+out_free_identity:
+	free(psk_identity);
+
+	return key;
+}
+
+long nvme_insert_tls_key(const char *keyring, const char *hostnqn,
+			const char *subsysnqn, int hmac,
+			unsigned char *configured_key, int key_len)
+{
+	const EVP_MD *md;
+	key_serial_t keyring_id, key;
+	unsigned char *retained_key;
+
+	switch (hmac) {
+	case 1:
+		md = EVP_sha256();
+		break;
+	case 2:
+		md = EVP_sha384();
+		break;
+	default:
+		fprintf(stderr, "Invalid key hash %df\n", hmac);
+		errno = EINVAL;
+		return -1;
+	}
+	keyring_id = find_key_by_type_and_desc("keyring", keyring, 0);
+	if (keyring_id < 0) {
+		fprintf(stderr, "keyring '%s' not available\n",
+			keyring);
+		errno = EINVAL;
+		return -1;
+	}
+	retained_key = derive_retained_key(md, hostnqn,
+					   configured_key, key_len);
+	if (!retained_key) {
+		fprintf(stderr, "Failed to derive retained key\n");
+		errno = EINVAL;
+		return -1;
+	}
+	key = derive_tls_key(md, keyring_id, hostnqn,
+			     subsysnqn, retained_key, key_len);
+	if (key < 0) {
+		fprintf(stderr, "Failed to derive TLS key\n");
+		errno = (int)key;
```

Comment:
Not sure about returning anything else than a E code in errno is really good idea in IMO.

Reply:
That's on purpose; I want to return the key serial from nvme_insert_tls_key() such that the caller knows which key has been generated (if he cares to look). So I'd rather change the convention to return '0' if we failed, and the actual key serial number on success.